include/genomics/printer.hpp

```diff
@@ -92,67 +95,97 @@ namespace genomics {
       return -(delim + 1);
     }
 
-    std::string off_target_string(genome_structure gs,
-                                  std::vector<std::vector<std::tuple<int64_t, match>>>& off_targets,
-                                  int64_t max_off_targets) {
+    float calculate_cfd(std::string sgRNA, std::string sequence, std::string PAM) {
+      float cfd = 1.0;
+      for (int i=0; i<sgRNA.length(); i++) {
+        char _sgRNA = sgRNA.at(i);
+        char _sequence = sequence.at(i);
+        if (_sgRNA != _sequence) {
+          if (_sgRNA == 'T') _sgRNA = 'U';
+          auto key = fmt::format("r{}:d{},{}", _sgRNA, static_cast<char>(toupper(complement(_sequence))), i+1);
+          cfd *= genomics::mm_scores[key];
+        }
+      }
+      cfd *= genomics::pam_scores[PAM.substr(1,2)];
+      return cfd;
+    }
+
+    std::tuple<std::string, float> off_target_fields(const kmer& k,
+                                                     genome_structure gs,
+                                                     std::vector<std::vector<std::tuple<int64_t, match>>>& off_targets,
+                                                     int64_t max_off_targets) {
       uint64_t delim = get_delim(gs);
-      std::string out("");
+      std::string offtarget_hex("");
+      float cfd_sum = 0.0;
 
-      for (uint64_t k = 0; k < off_targets.size(); k++) {
+      for (uint64_t i = 0; i < off_targets.size(); i++) {
           std::list<int64_t> v;
 
-          if (max_off_targets != -1) { 
-              std::shuffle(off_targets[k].begin(), off_targets[k].end(), std::mt19937{std::random_device{}()});
-              int64_t count = 0;
-              for (const auto& t : off_targets[k]) {
-                  if (count >= max_off_targets) break;
-                  v.push_back(std::get<0>(t));
-                  count++;
-              }
-          } else {
-            for (const auto& t : off_targets[k]) {
-                v.push_back(std::get<0>(t));
-            }
+          std::shuffle(off_targets[i].begin(), off_targets[i].end(), std::mt19937{std::random_device{}()});
```

Comment:
I had to comment out this line for the entire duration of my development to produce reproducible results. I've left it intact here, but maybe we remove it unless we absolutely need it?

Reply:
We do need this so that we don't bias towards any particular off-targets when we limit it. To make it reproducible we just need to set a random seed. Could add this as a CLI flag.